aggregator/flush_mgr.go

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errFlushManagerClosed = errors.New("flush manager is closed")
+)
+
+// PeriodicFlusher flushes periodically.
+type PeriodicFlusher interface {
+	// FlushInterval returns the periodic flush interval.
+	FlushInterval() time.Duration
+
+	// Flush performs a flush.
+	Flush()
+}
+
+// FlushManager manages and coordinates flushing activities across many
+// periodic flushers with different flush intervals with controlled concurrency
+// for flushes to minimize spikes in CPU load and reduce p99 flush latencies.
+type FlushManager interface {
+	// Register registers a metric list with the flush manager.
+	Register(flusher PeriodicFlusher) error
+
+	// Close closes the flush manager.
+	Close()
+}
+
+type flushManagerMetrics struct {
+	queueSize tally.Gauge
+}
+
+func newFlushManagerMetrics(scope tally.Scope) flushManagerMetrics {
+	return flushManagerMetrics{
+		queueSize: scope.Gauge("queue-size"),
+	}
+}
+
+type flushManager struct {
+	sync.Mutex
+
+	nowFn         clock.NowFn
+	checkEvery    time.Duration
+	jitterEnabled bool
+	workers       xsync.WorkerPool
+	scope         tally.Scope
+
```

Comment:
I usually separate the internal fields from the arguments/options so I can keep track which are consts and which aren't